toss/fitness/fitness_function_utils.py

```diff
@@ -138,27 +139,19 @@ def compute_space_coverage(number_of_spacecrafts: int, spin_axis: np.ndarray, sp
         j = np.argmin(np.abs(theta[:, np.newaxis] - theta_points), axis=0) # indices along theta axis
         k = np.argmin(np.abs(phi[:, np.newaxis] - phi_points), axis=0) # indices along phi axis
 
-        # Create a new boolean tensor corresponding to the previous visits.
+        # Create a new boolean tensor corresponding to candidate trajectory.
         new_tensor = np.full((len(r), len(theta), len(phi)), False)
-        indices_previous_trajectory  =  np.where(bool_tensor == True)
-        new_tensor[indices_previous_trajectory[0], indices_previous_trajectory[1], indices_previous_trajectory[2]] = True 
-        
-        # Update the new tensor with the proposed candidate trajectory.
         new_tensor[i, j, k] = True
 
-        # Identify number of unique visits in the region of interest
-        indices_unique_visits = np.where(new_tensor == True)
-        n_unique_visits = indices_unique_visits[0].shape[0]
+        # Update the new tensor according to previously spacecraft trajectories to remove overlap in the new candidate trajectory.
+        indices_previous_visits  =  np.where(bool_tensor == True)
+        new_tensor[indices_previous_visits[0], indices_previous_visits[1], indices_previous_visits[2]] = False 
 
-        # Compute ratio of uniquely visited regions to the number of regions defined by the tensor.
-        ratio_unique_visits = n_unique_visits / bool_tensor.size
-
-        # Get weights corresponding to the uniquely visited regions. 
-        r_idx = indices_unique_visits[0] #np.concatenate((previous_visits[0], new_visits[0])) # unique_visits[0]
-        sum_of_weights = np.sum(1/r[r_idx])
+        # Identify number of new visits in the region of interest
+        indices_new_visits = np.where(new_tensor == True)
```

Comment:
It seems, before this code was quite different as you were settings the values in `new_tensor` to True. Why False now?
(in old code)
```
new_tensor[indices_previous_trajectory[0], indices_previous_trajectory[1], indices_previous_trajectory[2]] = True
```

Btw, what is the purpose of

`indices_previous_visits = np.where(bool_tensor == True)`

Isn't that just 
`indices_previous_visits = np.where(bool_tensor)`
which is
`indices_previous_visits = bool_tensor` ?!

Also

```
new_tensor[indices_previous_visits[0], indices_previous_visits[1], indices_previous_visits[2]] = False
```

isn't this just new_tensor = np.logical_xor(new_tensor,indices_previous_visits) (in case you actually want False here, else it logical_or / and I guess?) that should be much, much fast as it uses broadcasting and no indices.

Reply:
My initial idea was to create a tensor (with the same dimension as bool_tensor) but now instead with normalized weights rather than boolean values. Subsequently, I also wanted to redefine the bool_tensor to be of 1:s and 0:s instead. However, since I couldn't come up with any computationally efficient and good way of doing a form of "dot product" between the 3D bool_tensor and a 3D weight_tensor, representing the score formula:
            Score = Sum(B_i * w_i),
I ended up continuing with the current False/True tensors instead. 

In detail, I create a new tensor only representing the current candidate trajectory being evaluated (i.e we set true for visited points by the candidate trajectory and false otherwise). Since re-visits should not result in new gain, we compare this "candidate tensor" with the bool_tensor of already visited points and remove all of the re-visits by assigning False to such points on the candidate tensor. 

Since the weight of a point i on the tensor is defined by 1/r_i, we can create a (N,1) vector consisting of every possible weight, I refer to to this vector as tensor_weights. Then, if we identify the index for each radial component corresponding to the visited points on the candidate tensor, we can use these indices to obtain a vector of all active weights and then just sum these up to obtain the score. 

I have now updated the comments and name for each tensor and variable to clarify this in the code.

Reply:
However, if you happen to know any fast way of doing such a dot product, that would probably be better and more interpretable since we in that case can remove all of the comparisons and reassign a 0-weight in the weight_tensor for every visited point.

Reply:
Something like this?

```py
# Let this be the weights, every entry [0,1], sums to 1
weights = np.array(tensor_shape,dtype="float")
# Let this be bool tensor of current trajectory, True where visited, false else
bool_tensor_current_trajectory = np.array(tensor_shape,dtype="bool")
# Let this be combined bool tensor of previous trajectories, True where visited, false else
bool_tensor_already_visited = np.array(tensor_shape,dtype="bool")

# We can get new visits with logic operators
previously_visited_in_current = np.logical_and(bool_tensor_current_trajectory,bool_tensor_already_visited)
new_visits_tensor = np.logical_xor(bool_tensor_current_trajectory,previously_visited_in_current)

# Then can just cast bool to float
score = np.multiply(weights, new_visits_tensor.as_type("float")).sum()

# Get total visited
bool_tensor_all_visited = np.logical_or(bool_tensor_current_trajectory,bool_tensor_already_visited)

return score, bool_tensor_all_visited
```

Reply:
Perfect! I've now added a weight_tensor of the same shape as bool_tensor, but with the corresponding normalized weights instead. Since we only require the part of the candidate_tensor that is relevant for the trajectory, I also reduced the size of each tensor used to compute the score, which will hopefully speed up the comparisons even further.